src/test/scala/com/hortonworks/spark/sql/hive/llap/TestLlapQueryExecutionListener.scala

```diff
@@ -0,0 +1,137 @@
+import org.apache.spark.sql.hive.llap.DefaultSource
+import org.apache.spark.sql.sources.{BaseRelation, Filter}
+import org.apache.spark.sql.types.StructType
+
+// scalastyle:off
+import org.scalatest.{BeforeAndAfterAll, BeforeAndAfterEach, FunSuite}
+// scalastyle:on
+
+class TestLlapQueryExecutionListener
+    // scalastyle:off funsuite
+    extends FunSuite with BeforeAndAfterAll with BeforeAndAfterEach {
+    // scalastyle:on funsuite
+
+  private var spark: SparkSession = _
+
+  override def beforeAll(): Unit = {
+    super.beforeAll()
+    spark = SparkSession.builder()
+      .master("local[2]")
+      .appName("LlapQueryExecutionListener")
+      .config(
+        "spark.sql.queryExecutionListeners",
+        classOf[LlapQueryExecutionListener].getCanonicalName)
+      .getOrCreate()
+  }
+
+  override def afterAll(): Unit = {
+    try {
+      spark.stop()
+    } finally {
+      super.afterAll()
+    }
+  }
+
+  override def afterEach(): Unit = {
+    try {
+      CloseCalls.clear()
+    } finally {
+      super.afterEach()
+    }
+  }
+
+  test("Closes all LlapRelations after query executions - basic") {
+    val df = spark.read.format(classOf[FakeDefaultSource].getCanonicalName).load()
+    assert(CloseCalls.closeCalls == 0)
+    df.collect()
+    assert(CloseCalls.closeCalls == 1, "Closing LlapRelation was not attempted.")
+  }
+
+  test("Closes all LlapRelations after query executions - union") {
+    val df1 = spark.read.format(classOf[FakeDefaultSource].getCanonicalName).load()
+    val df2 = spark.read.format(classOf[FakeDefaultSource].getCanonicalName).load()
+    val df3 = spark.read.format(classOf[FakeDefaultSource].getCanonicalName).load()
+    val unionDF = df1.union(df2).union(df3)
+    assert(CloseCalls.closeCalls == 0)
+    unionDF.collect()
+    assert(
+      CloseCalls.closeCalls == 3,
+      s"Closing LlapRelation should be attempted 2 but got ${CloseCalls.closeCalls}.")
+  }
+
+  test("Closes all LlapRelations after query executions - different sources") {
+    val df1 = spark.range(0, 10).toDF
+    val df2 = spark.read.format(classOf[FakeDefaultSource].getCanonicalName).load()
+    val unionDF = df1.union(df1).union(df2)
+    assert(CloseCalls.closeCalls == 0)
+    unionDF.show(0)
+    assert(
+      CloseCalls.closeCalls == 1,
+      s"Closing LlapRelation should be attempted 1 but got ${CloseCalls.closeCalls}.")
+  }
+
+  test("Closes all LlapRelations after query executions - SQL") {
+    spark.sql(s"""
+      |CREATE TEMPORARY TABLE tableA
+      |USING ${classOf[FakeDefaultSource].getCanonicalName}
+    """.stripMargin.replaceAll("\n", " "))
+    val df = spark.sql("SELECT * FROM tableA")
+    assert(CloseCalls.closeCalls == 0)
+    df.count()
+    assert(
+      CloseCalls.closeCalls == 1,
+      s"Closing LlapRelation should be attempted 1 but got ${CloseCalls.closeCalls}.")
+  }
+}
+
+/**
+ * It holds a variable to count the actual 'close' call. It should manually be cleared.
+ */
+object CloseCalls {
```

Comment:
I'm wondering if this is thread safe because we use `local[2]` here. Shall we add synchronization?

Reply:
Nice! will address other comments too soon.